Sources/NIO/Codec.swift

```diff
@@ -96,64 +106,95 @@ private extension ChannelHandlerContext {
 
 extension ByteToMessageDecoder {
 
+    /// Decode in a loop until there is nothing more to decode.
+    private func decodeLoop(ctx: ChannelHandlerContext, decodeFunc: (ChannelHandlerContext, inout ByteBuffer) throws -> DecodingState) throws {
+        while var slice = self.cumulationBuffer?.slice(), slice.readableBytes > 0 {
+            // Needed to later calculate how much we need to advance the readerIndex of the cumulationBuffer.
+            let readable = slice.readableBytes
+
+            // We fetch the writerIndex of the cumulationBuffer to make a good guess about if the cumulationBuffer changed in between due re-entrant call to
+            // channelRead after we called decodeFunc(...).
+            let writerIndex = self.cumulationBuffer!.writerIndex
+            let result = try decodeFunc(ctx, &slice)
+
+            guard self.cumulationBuffer != nil else {
+                // The cumulationBuffer was set to nil by either removing the decoder or closing the channel, just break the loop.
+                break
+            }
+
+            self.cumulationBuffer!.moveReaderIndex(forwardBy: readable - slice.readableBytes)
+
+            // If the user told us more data is needed we also need to ensure the writerIndex did not change in between.
+            // If the writerIndex changed we need to retry as there is more data delivered via re-entrance maybe.
+            if result == .needMoreData && self.cumulationBuffer!.writerIndex == writerIndex {
+                break
+            }
+        }
+    }
+
     /// Calls `decode` until there is nothing left to decode.
     public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-        var buffer = self.unwrapInboundIn(data)
+        // Either merge the received data into the existing cumulationBuffer or use it as the cumulationBuffer if none exists yet.
         if self.cumulationBuffer != nil {
+            var buffer = self.unwrapInboundIn(data)
             self.cumulationBuffer!.write(buffer: &buffer)
-            buffer = self.cumulationBuffer!
         } else {
-            self.cumulationBuffer = buffer
+            self.cumulationBuffer = self.unwrapInboundIn(data)
         }
 
         ctx.withThrowingToFireErrorAndClose {
-            // Running decode method until either the user returned `.needMoreData` or an error occurred.
-            while try decode(ctx: ctx, buffer: &buffer) == .`continue` && buffer.readableBytes > 0 { }
+            try self.decodeLoop(ctx: ctx, decodeFunc: self.decode)
         }
 
-        if buffer.readableBytes > 0 {
-            if self.shouldReclaimBytes(buffer: buffer) {
-                buffer.discardReadBytes()
+        // Discard the cumulationBuffer or discard read bytes if needed.
+        guard self.cumulationBuffer != nil else {
```

Comment:
Any reason not to use `guard let` here?

Reply:
Because we want to act on the `self.cumulationBuffer` on the following lines and not on a different copy ?